### PR TITLE
Add the Data Availability Header to BlockID

### PR DESCRIPTION
**Pull request by evan-forbes**
## Add the Data Availability Header to BlockID attempt IV: A New Hope :sparkles: 

This is the first, and most substantial, portion of a series of 3-4 PRs that will result in the addition and use of the data availability header. There's still a some debugging to finish, but most of the work is completed. I planned on separating this PR further into three individual PRs, but combing into a single PR was necessary to compile, let alone test.

- [X] Add the DAH to proto files
- [X] QoL methods and functions such as `MinDataAvailabilityHeader`, and `block.BlockID`
- [X] Usage and enforcement of the minimum data availability header
- [x] Pass IPFS object to `store.BlockStore` upon node starting up (to be done in a separate PR)
- [x] Modify `store.BlockStore` to load and save block data via IPFS (to be done in a separate PR)

Closes: #XXX

### CODE REVIEW

File: blockchain/v0/reactor.go [hunk 1]
```diff
@@ -382,10 +382,17 @@ FOR_LOOP:
 				didProcessCh <- struct{}{}
 			}
 
+			first.DataHash = nil
+			first.Hash()
```
Comment (liamsi) [inline blockchain/v0/reactor.go:386]: Is setting first.DataHash to nil to trigger recomputing the DAHeader?

Comment (evan-forbes) [reply blockchain/v0/reactor.go:386]: That makes a lot of sense! This was just for debugging purposes, as during validation, an error is thrown because the data availability header's `hash` field is nil due to it not being exported or in [equivalent proto type](https://github.com/lazyledger/lazyledger-core/blob/50f722a510dd2ba8e3d31931c9d83132d6318d4b/proto/tendermint/types/types.proto#L148). 

Do you think we should make a method to force the computation of the hash, or a method to copy the exported `header.DataHash` field into the `DataAvailabilityHeader.hash` field?

File: state/state.go [hunk 2]
```diff
@@ -188,10 +188,17 @@ func StateFromProto(pb *tmstate.State) (*State, error) { //nolint:golint
...
+	var bi *types.BlockID
+	if pb.LastBlockID.DataAvailabilityHeader != nil {
+		var err error
+		bi, err = types.BlockIDFromProto(&pb.LastBlockID)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		bi = &types.BlockID{}
 	}
```
Comment (evan-forbes) [inline state/state.go:200]: There are quite few errors reading empty BlockIDs from the dataStore, and so far, this is solution I've been using to patch them up. I think (hope) this is the main source of many of the remaining bugs.

Comment (evan-forbes) [reply state/state.go:200]: They're also getting silenced in some cases, which is leading to timeouts.

File: types/block.go [hunk 3]
```diff
@@ -126,20 +128,93 @@ func (dah *DataAvailabilityHeader) Hash() []byte {
...
+	return len(dah.ColumnRoots) == 0 || len(dah.RowsRoots) == 0
+}
+
+// MinDataAvailabilityHeader returns a hard coded copy of a data availability
+// header from empty block data
+func MinDataAvailabilityHeader() *DataAvailabilityHeader {
+	first, err := namespace.IntervalDigestFromBytes(
+		NamespaceSize,
+		hexBytesFromString(
+			"fffffffffffffffefffffffffffffffe669aa8f0d85221a05b6f0917884d30616a6c7d5330a5640a08a04dcc5b092f4f",
```
Comment (liamsi) [inline types/block.go:166]: If we hardcode this, why do we encode it as hex and decode it each time `MinDataAvailabilityHeader` is called?

File: types/block.go [hunk 4]
```diff
@@ -126,23 +128,100 @@ func (dah *DataAvailabilityHeader) Hash() []byte {
...
+func (dah *DataAvailabilityHeader) IsZero() bool {
+	if dah == nil {
+		return true
+	}
+	return len(dah.ColumnRoots) == 0 || len(dah.RowsRoots) == 0
+}
+
+// MinDataAvailabilityHeader returns a hard coded copy of a data availability
+// header from empty block data
+func MinDataAvailabilityHeader() *DataAvailabilityHeader {
```
Comment (liamsi) [inline types/block.go:162]: Can you also update the comment slightly: https://github.com/lazyledger/lazyledger-core/blob/d865cc7c917114f36b1d56e727c1a35c38d68c74/types/block.go#L312-L316 
(hash(empty) can be still misunderstood as hash(nil))

Comment (evan-forbes) [reply types/block.go:162]: this code is not really reachable, as `ComputeShares` will always return at least 1 share. While not the most explicit, `ComputeShares` returning at least a single share is also currently tested in [TestEmptyBlockData](https://github.com/lazyledger/lazyledger-core/blob/5c39da76bbc461074c97ed29585f1c549b6cca5f/types/block_test.go#L237). Therefore, I decided to remove this code entirely https://github.com/lazyledger/lazyledger-core/pull/312/commits/8acce6bab542afb6149a28e62b4fc9764585943c

happy to reverse that decision, though, if anyone thinks otherwise

Comment (evan-forbes) [reply types/block.go:162]: also added better docs to `ComputeShares` to reflect that it always returns at least a single share https://github.com/lazyledger/lazyledger-core/pull/312/commits/304537fb9e2c141e4f2e78c1196d979e617d4ee4

File: consensus/state_test.go [hunk 5]
```diff
@@ -444,6 +444,12 @@ func TestStateLockNoPOL(t *testing.T) {
 	vs2 := vss[1]
 	height, round := cs1.Height, cs1.Round
 
+	// timeouts have been increased because the CI keeps failing when the tests
+	// pass locally
+	precommitTimeout := cs1.config.Precommit(round).Nanoseconds() * 4
+	proposeTimeout := cs1.config.Propose(round).Nanoseconds() * 4
+	prevoteTimeout := cs1.config.Prevote(round).Nanoseconds() * 4
```
Comment (evan-forbes) [inline consensus/state_test.go:451]: @liamsi how terrible is this?

File: types/proposal.go [hunk 6]
```diff
@@ -35,15 +35,15 @@ type Proposal struct {
...
 		Type:      tmproto.ProposalType,
 		Height:    height,
 		Round:     round,
 		BlockID:   blockID,
 		POLRound:  polRound,
 		Timestamp: tmtime.Now(),
-		DAHeader:  daH,
+		DAHeader:  blockID.DataAvailabilityHeader,
 	}
 }
```
Comment (evan-forbes) [inline types/proposal.go:48]: The `Proposal` type now has two references to the data availability header, but that will change when we remove the block id completely.

**Comment by evan-forbes** [comment]
Marking this ready for review. There are still few tests that the CI doesn't like, but it looks like we can fix those by increasing timeouts.

File: types/block_test.go [hunk 7]
```diff
@@ -458,84 +466,82 @@ func TestMaxHeaderBytes(t *testing.T) {
...
-			}, "#%v", i)
-		} else {
-			assert.Equal(t,
-				tc.result,
-				MaxDataBytesNoEvidence(tc.maxBytes, tc.valsCount),
-				"#%v", i)
-		}
-	}
-}
+// func TestBlockMaxDataBytes(t *testing.T) {
```
Comment (Wondertan) [inline types/block_test.go:485]: What's the status of the tests below?

File: types/block_test.go [hunk 8]
```diff
@@ -289,50 +293,51 @@ func TestCommitValidateBasic(t *testing.T) {
 	}
 }
 
-func TestMaxCommitBytes(t *testing.T) {
-	// time is varint encoded so need to pick the max.
-	// year int, month Month, day, hour, min, sec, nsec int, loc *Location
-	timestamp := time.Date(math.MaxInt64, 0, 0, 0, 0, 0, math.MaxInt64, time.UTC)
+// func TestMaxCommitBytes(t *testing.T) {
```
Comment (Wondertan) [inline types/block_test.go:296]: And this one

File: store/store_test.go [hunk 9]
```diff
@@ -382,11 +392,14 @@ func TestLoadBaseMeta(t *testing.T) {
...
+	lastCommit := &types.Commit{BlockID: types.EmptyBlockID()}
+
 	for h := int64(1); h <= 10; h++ {
-		block := makeBlock(h, state, new(types.Commit))
+		block := makeBlock(h, state, lastCommit)
 		partSet := block.MakePartSet(2)
-		seenCommit := makeTestCommit(h, tmtime.Now())
-		bs.SaveBlock(block, partSet, seenCommit)
+		lastCommit := makeTestCommit(block, h, tmtime.Now())
+		block.Header.LastBlockID = lastCommit.BlockID
```
Comment (Wondertan) [inline store/store_test.go:401]: Maybe move this line into `makeBlock`?

Comment (Wondertan) [reply store/store_test.go:401]: This would simplify some code here and below

File: state/state.go [hunk 10]
```diff
@@ -188,10 +188,17 @@ func StateFromProto(pb *tmstate.State) (*State, error) { //nolint:golint
 	state.ChainID = pb.ChainID
 	state.InitialHeight = pb.InitialHeight
 
-	bi, err := types.BlockIDFromProto(&pb.LastBlockID)
-	if err != nil {
-		return nil, err
+	var bi *types.BlockID
+	if pb.LastBlockID.DataAvailabilityHeader != nil {
```
Comment (Wondertan) [inline state/state.go:192]: Moving this check down to BlockIDFromProto would make the code cleaner

File: node/node_test.go [hunk 11]
```diff
@@ -395,7 +400,7 @@ func TestMaxProposalBlockSize(t *testing.T) {
 
 	state, stateDB, _ := state(types.MaxVotesCount, int64(1))
 	stateStore := sm.NewStore(stateDB)
-	const maxBytes int64 = 1024 * 1024 * 2
+	const maxBytes int64 = 3151945
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply node/node_test.go:403]: great question, by simply using the number provided by the failing test :sweat_smile: 
```sh
--- FAIL: TestMaxProposalBlockSize (0.58s)
panic: Negative MaxDataBytesUnknownEvidence. Block.MaxBytes=2097152 is too small to accommodate header&lastCommit&evidence=3150945 [recovered]
        panic: Negative MaxDataBytesUnknownEvidence. Block.MaxBytes=2097152 is too small to accommodate header&lastCommit&evidence=3150945
```

I figured it wasn't super important, as anything related to testing for a max number of bytes will have to be changed and reevaluated after the blockID is fully removed, as the DataAvailabilityHeader will obviously not have a constant size

File: consensus/state.go [hunk 12]
```diff
@@ -1213,14 +1217,18 @@ func (cs *State) defaultDoPrevote(height int64, round int32) {
...
+			DataAvailabilityHeader: &cs.LockedBlock.DataAvailabilityHeader},
+		)
 		return
 	}
 
 	// If ProposalBlock is nil, prevote nil.
 	if cs.ProposalBlock == nil {
 		logger.Info("enterPrevote: ProposalBlock is nil")
-		cs.signAddVote(tmproto.PrevoteType, nil, types.PartSetHeader{})
+		cs.signAddVote(tmproto.PrevoteType, types.EmptyBlockID())
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply consensus/state.go:1231]: That's a good point. Personally, I think `EmptyBlockID` makes sense because its the block id for a block with no data, not a block id for a nil block. Definitely not attached to that name by any means, though.

File: proto/tendermint/types/canonical.proto [hunk 13]
```diff
@@ -10,6 +10,7 @@ import "google/protobuf/timestamp.proto";
 message CanonicalBlockID {
   bytes                  hash            = 1;
   CanonicalPartSetHeader part_set_header = 2 [(gogoproto.nullable) = false];
+  DataAvailabilityHeader data_availability_header = 3;
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply proto/tendermint/types/canonical.proto:13]: I like this idea, but if we decide to go through with it, I think we should handle it in #287

File: state/helpers_test.go [hunk 14]
```diff
@@ -54,6 +54,7 @@ func makeAndCommitGoodBlock(
 
 func makeAndApplyGoodBlock(state sm.State, height int64, lastCommit *types.Commit, proposerAddr []byte,
 	blockExec *sm.BlockExecutor, evidence []types.Evidence) (sm.State, types.BlockID, error) {
+	state.LastBlockID = lastCommit.BlockID
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply state/helpers_test.go:57]: Yeah, all data availability headers that are nil/empty throw an error during `ValidateBasic`, which is called when converting the state to and from protobuf.

File: types/proposal.go [hunk 15]
```diff
@@ -35,15 +35,15 @@ type Proposal struct {
...
+func NewProposal(height int64, round int32, polRound int32, blockID BlockID) *Proposal {
 	return &Proposal{
 		Type:      tmproto.ProposalType,
 		Height:    height,
 		Round:     round,
 		BlockID:   blockID,
 		POLRound:  polRound,
 		Timestamp: tmtime.Now(),
-		DAHeader:  daH,
+		DAHeader:  blockID.DataAvailabilityHeader,
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply types/proposal.go:46]: exactly

File: consensus/common_test.go [hunk 16]
```diff
@@ -52,7 +52,7 @@ type cleanupFunc func()
 var (
 	config                *cfg.Config // NOTE: must be reset for each _test.go file
 	consensusReplayConfig *cfg.Config
-	ensureTimeout         = 600 * time.Millisecond
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply consensus/common_test.go:55]: while the increase is not dialed to exactly what it needs to be, having this increase made the CI tests significantly more reliable.

File: state/store.go [hunk 17]
```diff
@@ -150,6 +150,8 @@ func (store dbStore) loadState(key []byte) (state State, err error) {
 	return *sm, nil
 }
 
+// The state cannot be unmarshalled because it contains some empty dah
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply state/store.go]: nope good catch, resolved here https://github.com/lazyledger/lazyledger-core/pull/312/commits/371684e1feaffa885ce278f75665d2e76dee2d14

File: store/store_test.go [hunk 18]
```diff
@@ -374,6 +382,8 @@ func TestBlockStoreSaveLoadBlock(t *testing.T) {
 	}
 }
 
+// the header and the blockID hashes are not matching during pruneblocks
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply store/store_test.go]: nope good catch, resolved here 371684e

File: types/evidence_test.go [hunk 19]
```diff
@@ -265,28 +266,12 @@ func makeHeaderRandom() *Header {
 func TestEvidenceProto(t *testing.T) {
 	// -------- Votes --------
 	val := NewMockPV()
-	blockID := makeBlockID(tmhash.Sum([]byte("blockhash")), math.MaxInt32, tmhash.Sum([]byte("partshash")))
-	blockID2 := makeBlockID(tmhash.Sum([]byte("blockhash2")), math.MaxInt32, tmhash.Sum([]byte("partshash")))
+	blockID := EmptyBlockID()
+	blockID2 := EmptyBlockID()
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply types/evidence_test.go]: good point, changed in https://github.com/lazyledger/lazyledger-core/pull/312/commits/b28f6cb0fd9c46650edcc52e0d9eaf3a58d6ad40

File: types/block_test.go (hunk 7 above)
Comment (evan-forbes) [reply types/block_test.go:485]: this is another test that I commented out rather than fix, as checks for max bytes should be reevaluated. 

I definitely need to keep track of this in an issue if we decide to keep these commented out.

Curious as to what @liamsi thinks.

File: store/store_test.go (hunk 9 above)
Comment (evan-forbes) [reply store/store_test.go]: It would make things simpler here, but we'd have to add logic to use an empty block id when the commit is nil, and lose some explicitness. I did however remove an unnecessary line https://github.com/lazyledger/lazyledger-core/pull/312/commits/17055a67fc101d161fde0828729b4b8972592ac3

File: state/state.go (hunk 10 above)
Comment (evan-forbes) [reply state/state.go:192]: ahh yes, but then BlockIDFromProto would do more than just convert to and from proto, and that felt like a leaky abstraction waiting to happen so decided against that.

File: consensus/state.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (Wondertan) [reply consensus/state.go:1231]: Yeah, from this perspective it also makes sense, so here we have a semantical meaning tradeoff or something. I am ok with any option btw)

File: consensus/state_test.go (hunk 5 above)
Comment (Wondertan) [reply consensus/state_test.go:451]: @evan-forbes?

File: node/node_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (Wondertan) [reply node/node_test.go:403]: Makes sense, thanks.

File: proto/tendermint/types/canonical.proto (hunk 13 above)
[the remark this answers is not in the document]
Comment (Wondertan) [reply proto/tendermint/types/canonical.proto:13]: Okkk, but why not now if we introducing the field here? Anyway, I am not insisting on that.

File: types/block_test.go (hunk 7 above)
Comment (Wondertan) [reply types/block_test.go:485]: @adlerjohn, I feel like I asked this before, but anyway it seems like we need to superficially discuss the following again. Are there any constraints for max block size?

File: state/state.go (hunk 10 above)
Comment (Wondertan) [reply state/state.go:192]: Ok

File: types/block_test.go (hunk 7 above)
Comment (adlerjohn) [reply types/block_test.go:485]: The constraint on maximum block size is whatever the erasure coding library supports. I have to look into exactly what it is, but at the very least a 128*128 original square with 256 bytes per share should work.

**Review by Wondertan — APPROVED** [review]
Besides minor unresolved comments and in advance of other possible small changes @evan-forbes may decide to introduce here, this looks really good to me. Thanks for this important, though unpleasant and cumbersome to implement, step further 👍🏻

File: proto/tendermint/types/canonical.proto (hunk 13 above)
[the remark this answers is not in the document]
Comment (evan-forbes) [reply proto/tendermint/types/canonical.proto:13]: I mean, we're going to remove blockID and CanonicalBlockID completely in #287, and then the remaining data availability header fields can all be changed to the short hand version consistently across the codebase.

[the remark this answers is not in the document]
Comment (evan-forbes) [reply proto/tendermint/types/canonical.proto:13]: The only place that uses shorthand now is things involving `Proposal`, everything else uses the long form.

File: consensus/state_test.go (hunk 5 above)
Comment (evan-forbes) [reply consensus/state_test.go:451]: It could be that the timeouts are so short, (10000000ns, 40000000ns, 10000000ns respectively) that adding any computation in a resource constrained env (github actions) that this is necessary. Tests pass locally with out the additional time, but tbh I'm not 100% sure

Comment (Wondertan) [reply consensus/state_test.go:451]: Yeah, but they passed previously on CI, and now they are not, though the PR does not introduce any additional computational requirements AFAIU. That is why this does not make sense to me.

File: proto/tendermint/types/canonical.proto (hunk 13 above)
[the remark this answers is not in the document]
Comment (Wondertan) [reply proto/tendermint/types/canonical.proto:13]: 👌

File: types/block_test.go (hunk 7 above)
Comment (evan-forbes) [reply types/block_test.go:485]: While I think everything testing for Maximum bytes of X will be easier after #287, as we will not have to account for the max data availability header, this particular case is strongly related to #77

File: consensus/state_test.go (hunk 5 above)
Comment (evan-forbes) [reply consensus/state_test.go:451]: I'm still not 100% positive that this is the cause, but there is more computation introduced each time the [blockID is validated](https://github.com/lazyledger/lazyledger-core/blob/17055a67fc101d161fde0828729b4b8972592ac3/types/block.go#L1772-L1772). Before their was no obviously no data availability header, and now a minimum data availability header has to be hashed if it hasn't already.

Comment (evan-forbes) [reply consensus/state_test.go:451]: and [hashing the header during validation](https://github.com/lazyledger/lazyledger-core/blob/17055a67fc101d161fde0828729b4b8972592ac3/types/block.go#L145-L147) might be a mistake.

Comment (evan-forbes) [reply consensus/state_test.go:451]: > and hashing the header during validation might be a mistake.

ahh ohkay, I remember now, the dah.hash field is not exported, so converting to and from proto erases it and makes a bunch of tests angry, so I added that extra hash command during validation. For the record, I really don't like this. Validating the struct should obviously not modify the struct by adding the hash. For now though, unless we want to export the hash field in `DataAvailabilityHeader`, we have to do this.

Do we want to export the hash field?

Comment (liamsi) [reply consensus/state_test.go:451]: Hmm, exporting the field would mean that we introduce something that is not spec compliant (cc @adlerjohn). the hash is something that is currently only used internally for some kind of caching. So modifying that field should happen with a call to dah.Hash() and it is only to prevent that the hash is recomputed each time. It does not necessarily need to happen during validation though.

Comment (liamsi) [reply consensus/state_test.go:451]: Instead of cs1.config.Precommit(round).Nanoseconds() * 4, you could modify the config for that test.

File: types/block_test.go (hunk 7 above)
Comment (liamsi) [reply types/block_test.go:485]: > Curious as to what @liamsi thinks.

I think it should not be too difficult to figure out the numbers but it is OK to do this in the followup pr as you said (once instead twice). And yes, best to track it in an issue to make sure we won't forget 👍🏼

File: consensus/state_test.go (hunk 5 above)
Comment (evan-forbes) [reply consensus/state_test.go:451]: > So modifying that field should happen with a call to dah.Hash() and it is only to prevent that the hash is recomputed each time. It does not necessarily need to happen during validation though.

definitely agree. The only problem arises when considering tests such as `TestMsgToProto`. It is comparing proto messages before and after encoding/decoding, during which the unexported hash field gets erased. In this case, if we want to call `dah.Hash()` we'd have to type check each message https://github.com/lazyledger/lazyledger-core/blob/17055a67fc101d161fde0828729b4b8972592ac3/consensus/msgs_test.go#L23-L221

Does this change things at all? Should we keep filling in the hash during validation?

If so, I think we can remove it come #278.

Comment (liamsi) [reply consensus/state_test.go:451]: I didn't fully look into that test but if it is just a proto concern, why not fill that field when doing the *FromProto instead?

Comment (evan-forbes) [reply consensus/state_test.go]: that seems like the best solution! https://github.com/lazyledger/lazyledger-core/pull/312/commits/10366d2b6f7f01bda263e96a82333a41f9d854ba was even able to remove the timeouts! https://github.com/lazyledger/lazyledger-core/pull/312/commits/9fd4f9f86e8a7c2f633d6f546a898eb852a95ac8

thanks, @liamsi

Comment (Wondertan) [reply consensus/state_test.go]: Noice👌

File: types/block.go [hunk 20]
```diff
@@ -1649,6 +1743,9 @@ func (blockID BlockID) ValidateBasic() error {
 	if err := blockID.PartSetHeader.ValidateBasic(); err != nil {
 		return fmt.Errorf("wrong PartSetHeader: %v", err)
 	}
+	if err := blockID.DataAvailabilityHeader.ValidateBasic(); err != nil {
+		return fmt.Errorf("wrong DataAvailabilityHeader: %w", err)
+	}
```
Comment (liamsi) [inline types/block.go:1748]: 👍🏼

File: types/block.go [hunk 21]
```diff
@@ -1682,16 +1779,17 @@ func (blockID *BlockID) ToProto() tmproto.BlockID {
...
+		DataAvailabilityHeader: blockID.DataAvailabilityHeader.ToProto(),
 	}
 }
 
 // FromProto sets a protobuf BlockID to the given pointer.
 // It returns an error if the block id is invalid.
 func BlockIDFromProto(bID *tmproto.BlockID) (*BlockID, error) {
 	if bID == nil {
-		return nil, errors.New("nil BlockID")
+		return nil, nil
```
Comment (liamsi) [inline types/block.go:1792]: is this change really necessary?

File: store/store_test.go [hunk 22]
```diff
@@ -33,7 +33,7 @@ import (
 type cleanupFunc func()
 
 // make a Commit with a single vote containing just the height and a timestamp
-func makeTestCommit(height int64, timestamp time.Time) *types.Commit {
+func makeTestCommit(block *types.Block, height int64, timestamp time.Time) *types.Commit {
```
Comment (liamsi) [inline store/store_test.go:36]: nit: seems a bit exessive to pass in the the whole block here.

**Review by liamsi — APPROVED** [review]
Awesome!!! One major step closer to actual DAS nodes (not only the light client) 👏🏼

**Comment by evan-forbes** [comment]
So, e2e tests pass locally, don't pass in CI. That lead me to think that it was timeout related, but now I don't think that's the case. [The logs](https://pipelines.actions.githubusercontent.com/1SA1FIc8M0EvkNoxfkLxZJsZcnYZrIs3EZFxQoCDkoJGDQq1Gi/_apis/pipelines/1/runs/5275/signedlogcontent/3?urlExpires=2021-05-21T02%3A22%3A21.2278303Z&urlSigningMethod=HMACV1&urlSignature=6xVwuRS7%2FRMFWgPvsarZjuOTPet81Q2uv1gRa5oROgk%3D) are quite different from logs of the last PR's e2e failure [logs](https://pipelines.actions.githubusercontent.com/1SA1FIc8M0EvkNoxfkLxZJsZcnYZrIs3EZFxQoCDkoJGDQq1Gi/_apis/pipelines/1/runs/5238/signedlogcontent/3?urlExpires=2021-05-21T02%3A21%3A36.8035502Z&urlSigningMethod=HMACV1&urlSignature=507EtfhUcV3InKCCZg%2Fygd9UnWWfg7aH9fAtMzkDxkg%3D)

will debug further tomorrow.

**Comment by liamsi** [comment]
I ran into similar problems with the e2e tests in https://github.com/lazyledger/lazyledger-core/pull/323. This change made them pass on CI as well: https://github.com/lazyledger/lazyledger-core/pull/323#discussion_r637504780 
Maybe that helps.

File: store/store_test.go (hunk 22 above)
Comment (evan-forbes) [reply store/store_test.go]: shoot, thought I responded to this. It is not necessary and has been removed https://github.com/lazyledger/lazyledger-core/pull/312/commits/3f1c2f2d5156373c02df815ca6c937b965378921

File: types/block.go (hunk 21 above)
Comment (evan-forbes) [reply types/block.go]: at one point while debugging, it was, but now it's not. Good catch, thanks! https://github.com/lazyledger/lazyledger-core/pull/312/commits/a8a81ee7dad4fc7ef89c4d09d35e3fd24cbdb00a

File: store/store_test.go (hunk 22 above)
Comment (liamsi) [reply store/store_test.go]: no worries. I've noticed that you've updated it already.

File: test/e2e/runner/wait.go [hunk 23]
```diff
@@ -30,5 +30,5 @@ func WaitUntil(testnet *e2e.Testnet, height int64) error {
 // waitingTime estimates how long it should take for a node to reach the height.
 // More nodes in a network implies we may expect a slower network and may have to wait longer.
 func waitingTime(nodes int) time.Duration {
-	return time.Duration(30+(nodes*4)) * time.Second
+	return time.Duration(30+(nodes*120)) * time.Second
```
Comment (liamsi) [inline test/e2e/runner/wait.go:33]: Seems a bit dramatic to multiply by 120. If this work reliably I'd suggest something like:
```suggestion
	return time.Duration(60+(nodes*6)) * time.Second
```
On the other hand, we have time I guess :D

**Comment by liamsi** [comment]
btw seems like increasing the ensure timeout (https://github.com/lazyledger/lazyledger-core/commit/fe558820b1dc824c1db91ac292952696c47c37d0) wasn't the smartest move either. Some of the consensus unit tests still fail. But this is orthogonal to the e2e tests.

File: test/e2e/runner/wait.go (hunk 23 above)
Comment (evan-forbes) [reply test/e2e/runner/wait.go:33]: I was just trying something insane to get the tests to pass. This worked on an identical branch here https://github.com/evan-forbes/lazyledger-core/pull/37. Those e2e pass consistently, for some reason these still don't.

**Comment by liamsi** [comment]
<details><summary>click here for snippet from e2e log</summary>

```
validator02    | E[2021-05-24|23:45:17.130] SignerServer: HandleMessage                  err="message exceeds max size (15532 > 10240) *privval.Message *privval.SecretConnection"
validator02    | E[2021-05-24|23:45:17.130] SignerServer: HandleMessage                  err="message exceeds max size (11351 > 10240) *privval.Message *privval.SecretConnection"
validator02    | E[2021-05-24|23:45:17.130] SignerServer: HandleMessage                  err="proto: wrong wireType = 5 for field SignedProposalResponse"
validator02    | E[2021-05-24|23:45:17.130] SignerServer: HandleMessage                  err="message exceeds max size (3108748918783 > 10240) *privval.Message *privval.SecretConnection"
validator02    | E[2021-05-24|23:45:17.130] SignerServer: HandleMessage                  err="proto: wrong wireType = 3 for field PubKeyResponse"
validator02    | E[2021-05-24|23:45:18.130] SignerServer: HandleMessage                  err="proto: wrong wireType = 3 for field PubKeyResponse"
validator02    | E[2021-05-24|23:45:18.130] SignerServer: HandleMessage                  err="message exceeds max size (128646886 > 10240) *privval.Message *privval.SecretConnection"
validator02    | E[2021-05-24|23:45:18.130] SignerServer: HandleMessage                  err="unexpected EOF"
validator02    | E[2021-05-24|23:45:18.130] SignerServer: HandleMessage                  err="unexpected EOF"
validator02    | E[2021-05-24|23:45:18.130] SignerServer: HandleMessage                  err="message exceeds max size (4998432636427829247 > 10240) *privval.Message *privval.SecretConnection"
validator02    | E[2021-05-24|23:45:18.130] SignerServer: HandleMessage                  err="message exceeds max size (698005 > 10240) *privval.Message *privval.SecretConnection"
validator02    | E[2021-05-24|23:45:18.130] SignerServer: HandleMessage                  err="message exceeds max size (11528 > 10240) *privval.Message *privval.SecretConnection"
validator02    | E[2021-05-24|23:45:18.130] SignerServer: HandleMessage                  err="proto: illegal wireType 6"
```
</details>

**Comment by evan-forbes** [comment]
So, I've updated this PR to master, and it's still really weird. The e2e tests pass fine locally, but when we run the e2e test in the CI, we get what looks like a protobuf encoding/decoding off by 1 error. 

Our needs have changed since the inception of this PR, as we no longer _necessarily_ need to replace the `PartSetHeader`, although it would be nice if would could. Should we decide to, then it might also make sense to merge this PR even with the CI being broken.

**Comment by liamsi** [comment]
> The e2e tests pass fine locally, but when we run the e2e test in the CI, we get what looks like a protobuf encoding/decoding off by 1 error.

Do the tests fail consistently on CI and always pass locally?

**Comment by evan-forbes** [comment]
> Do the tests fail consistently on CI and always pass locally?

yeah, it's really weird.

File: types/block.go [hunk 24]
```diff
@@ -100,9 +100,10 @@ func (dah *DataAvailabilityHeader) Equals(to *DataAvailabilityHeader) bool {
 
 // Hash computes and caches the merkle root of the row and column roots.
 func (dah *DataAvailabilityHeader) Hash() []byte {
-	if dah == nil {
-		return merkle.HashFromByteSlices(nil)
+	if dah.IsZero() {
+		dah = MinDataAvailabilityHeader()
 	}
```
Comment (Wondertan) [inline types/block.go:105]: @evan-forbes, I guess this is wrong. Zero DAH is when rows len and cols len are equal to zero, while MinDataAvailabilityHeader has 2 rows and 2 cols, as empty block data does not mean not data at all

Comment (Wondertan) [reply types/block.go:105]: I would just remove this if

Comment (Wondertan) [reply types/block.go:105]: Or the intuition here is to never actually use empty DAH and always swap it with empty DAH, right?

**Comment by Wondertan** [comment]
@liamsi, I guess we need your decision on how to move forward, or let's keep it open until it is clear on how to resolve BlockID situation. Anyway, we have here some nice changes that I would like to keep, like ValidateBasic for DAHeader and MinDataAvailabilityHeader.

**Comment by liamsi** [comment]
Let's merge #441 first, then update this work and look into how to move forward in combination with #427 .

**Review by liamsi — CHANGES_REQUESTED** [review]
Removing my approval s.t. we don't accidentally merge this before #441

**Comment by liamsi** [comment]
@evan-forbes are we OK to close this one?

**Comment by Wondertan** [comment]
Guess so, waiting for @evan-forbes

**Comment by evan-forbes** [comment]
sorry the late response, yeah definitely close this one
